python/ray/tune/utils/file_transfer.py

```diff
@@ -0,0 +1,234 @@
+    """Return dict of files mapping to stats in ``path``.
+
+    This function scans a directory ``path`` recursively and returns a dict
+    mapping each contained file to a tuple of (mtime, filesize).
+
+    mtime and filesize are returned from ``os.lstat`` and are usually a
+    floating point number (timestamp) and an int (filesize in bytes).
+    """
+    files_stats = {}
+    for root, dirs, files in os.walk(path, topdown=False):
+        rel_root = os.path.relpath(root, path)
+        for file in files:
+            key = os.path.join(rel_root, file)
+            stat = os.lstat(os.path.join(path, key))
+            files_stats[key] = stat.st_mtime, stat.st_size
+
+    return files_stats
+
+
+# Only export once
+_remote_get_recursive_files_and_stats = ray.remote(_get_recursive_files_and_stats)
+
+
+def _pack_dir(
+    source_dir: str, files_stats: Optional[Dict[str, Tuple[float, int]]] = None
+) -> io.BytesIO:
+    """Pack whole directory contents into an uncompressed tarfile.
+
+    This function accepts a ``files_stats`` argument. If given, only files
+    whose stats differ from these stats will be packed.
+
+    The main use case for this is that we can collect information about files
+    already existing in the target directory, and only pack files that have
+    been updated. This is similar to how cloud syncing utilities decide
+    which files to transfer.
+
+    Args:
+        source_dir: Path to local directory to pack into tarfile.
+        files_stats: Dict of relative filenames mapping to a tuple of
+            (mtime, filesize). Only files that differ from these stats
+            will be packed.
+
+    Returns:
+        Tarfile as a stream object.
+    """
+    stream = io.BytesIO()
+    with tarfile.open(fileobj=stream, mode="w", format=tarfile.PAX_FORMAT) as tar:
+        if not files_stats:
+            # If no `files_stats` is passed, pack whole directory
+            tar.add(source_dir, arcname="", recursive=True)
+        else:
+            # Otherwise, only pack differing files
+            tar.add(source_dir, arcname="", recursive=False)
+            for root, dirs, files in os.walk(source_dir, topdown=False):
+                rel_root = os.path.relpath(root, source_dir)
+                # Always add all directories
+                for dir in dirs:
+                    key = os.path.join(rel_root, dir)
+                    tar.add(os.path.join(source_dir, key), arcname=key, recursive=False)
+                # Add files where our information differs
+                for file in files:
+                    key = os.path.join(rel_root, file)
+                    stat = os.lstat(os.path.join(source_dir, key))
+                    file_stat = stat.st_mtime, stat.st_size
+                    if key not in files_stats or file_stat != files_stats[key]:
+                        tar.add(os.path.join(source_dir, key), arcname=key)
+
+    return stream
+
+
+@ray.remote
+class _PackActor:
```

Comment:
Thoughts on exposing a configurable `max_size` which can be used to warn or raise an exception if surpassed? (even though this isn't actually exposed to the user today)

Anecdotally I've appreciated runtime envs immediately erroring out when I accidentally include a 50GB dataset 😄 .

Reply:
I've added a max size bytes argument defaulting to 1 GB, but it is disabled in the sync client. I agree if users call this method they should be protected from these transfers (even if this is not a public API right now). in the sync client we mostly sync checkpoints and actually have to support large file sizes, so a default limit doesn't really make sense.